File: Cracking Coding Interview/Through Java/LearningToSolving/src/Course/Module_2/Assignment/Task_2.java
package Course.Module_2.Assignment;

import java.util.Scanner;

public class Task_2 {
    // Solution of the task 2
    public int solution(int[] prices) {
        int l = prices.length, tp = 0, mp = 0, j = 1;
        for (int i = 0; i < l-1; i++) {
            if (j >= l) {
                break;
            }
            else if (prices[i] < prices[j]) {
                tp = prices[j]-prices[i];
                if (tp > mp) {
                    mp = tp;
                }
                j++;
                i--;
            } else {
                tp = prices[j]-prices[i];
                if (tp > mp) {
                    mp = tp;
                }
                i = j-1;
                j++;
            }
        }

        return mp;
    }

    // Main method
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        System.out.print("Enter the size of the array: ");
        int size = sc.nextInt();

        int[] arr = new int[size];
        System.out.print("Enter the elements of the array: ");
        for (int i = 0; i < size; i++) {
            arr[i] = sc.nextInt();
        }

        Task_2 obj1 = new Task_2();
        System.out.println(obj1.solution(arr));
    }
}


// Time Complexity : O(N)
// Space Complexity : O(N)

File: Cracking Coding Interview/Through C++/Module-2/Assignment/Task_2.cpp
#include <bits/stdc++.h>

using namespace std;

class Task_2 {
	public:
		// Solve the task-2 assignment
		void solution(int prices[], int size) {
			int temp = 0, j = 0;

            for (int i = 1; i < size; i++) {
                if (prices[i]-prices[j] > temp) {
                    temp = prices[i]-prices[j];
                } else {
                    if (prices[i] > prices[j]) {
                        continue;
                    } else {
                        j = i;
                    }
                }
            }

            cout << temp << "\n";
		}
};


int main() {
    int size;
    cout << "Enter the size the day: ";
    cin >> size;

    int prices[size];
    cout << "Enter the prices list of the days: ";
    for (int i = 0 ; i < size; i++) {
        cin >> prices[i];
    }

	Task_2 obj;
	obj.solution(prices, size);

	return 0;
}


// Time complexity O(N)
// Space Complexity O(N)